newsroom/wire/formatters/utils.py

```diff
@@ -0,0 +1,140 @@
+    :param item:
+    :return:
+    """
+    added_files = []
+    for _key, associated_item in item.get('associations', {}).items():
+        for rendition in associated_item.get('renditions'):
+            name = associated_item.get('renditions').get(rendition).get('href').lstrip('/')
+            if name in added_files:
+                continue
+            file = flask.current_app.media.get(associated_item.get('renditions').get(rendition).get('media'),
+                                               ASSETS_RESOURCE)
+            zf.writestr(name, file.read())
+            added_files.append(name)
+
+
+def rewire_featuremedia(item):
+    """
+    Set the references in the feature media strip the leading / to make it a legitimate relative path
+    :param item:
+    :return:
+    """
+    renditions = item.get('associations', {}).get('featuremedia', {}).get('renditions', [])
+    for rendition in renditions:
+        item['associations']['featuremedia']['renditions'][rendition]['href'] = \
+            item['associations']['featuremedia']['renditions'][rendition]['href'].lstrip('/')
+
+
+def log_media_downloads(item):
+    """
+    Given an item create a download entry for all the associations
+    :param item:
+    :return:
+    """
+    for _key, associated_item in item.get('associations', {}).items():
+        action = 'download ' + associated_item.get('type')
+        get_resource_service('history').create_media_history_record(item, _key, action, get_user(required=True),
+                                                                    flask.request.args.get('type', 'wire'))
+
+
+def remove_unpermissioned_embeds(item, company_id=None, section='wire'):
+    """
+    :param item:
+    :param company_id:
+    :param section
+    :return: The item with the embeds that the user is not allowed to download removed
+    """
+
+    if not app.config.get("EMBED_PRODUCT_FILTERING"):
+        return
+
+    kill_keys = []
+
+    if company_id is None:
+        user = get_user(required=False)
+        if user:
+            company_id = user.get('company')
+        else:
+            company_id = flask.g.user
+
+    # get the list of superdesk products that the company is permissioned for
+    permitted_products = [p.get('sd_product_id') for p in
+                          get_products_by_company(company_id, None, section) if p.get('sd_product_id')]
+
+    for key, embed_item in item.get("associations", {}).items():
+        if key.startswith("editor_"):
+            # get the list of products that the embedded item matched in superdesk
+            embed_products = [p.get('code') for p in
+                              ((item.get('associations') or {}).get(key) or {}).get('products', [])]
+
+            if not len(set(embed_products) & set(permitted_products)):
+                kill_keys.append(key)
+
+    # Nothing to do
+    if len(kill_keys) == 0:
+        return
+
+    root_elem = lxml_html.fromstring(item.get('body_html', ''))
```

Comment:
The code to iterate over embeds is used multiple times, might be good to have a function that allows to provide a callback for each group found